src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs

```diff
@@ -16,44 +16,54 @@
 // 
 
 using System;
-using Nethermind.Blockchain.Find;
+using System.Runtime.CompilerServices;
 using Nethermind.Blockchain.Spec;
 using Nethermind.Core;
 using Nethermind.Core.Specs;
 using Nethermind.Int256;
 using Nethermind.State;
 using Nethermind.TxPool;
 
+[assembly:InternalsVisibleTo("Nethermind.TxPool.Test")]
+
 namespace Nethermind.Blockchain
 {
     public class ChainHeadInfoProvider : IChainHeadInfoProvider
     {
-        private readonly IBlockTree _blockTree;
-        
         public ChainHeadInfoProvider(ISpecProvider specProvider, IBlockTree blockTree, IStateReader stateReader)
             : this(new ChainHeadSpecProvider(specProvider, blockTree), blockTree, new ChainHeadReadOnlyStateProvider(blockTree, stateReader))
         {
         }
-        public ChainHeadInfoProvider(ISpecProvider specProvider, IBlockTree blockTree, IReadOnlyStateProvider stateProvider)
+        
+        public ChainHeadInfoProvider(ISpecProvider specProvider, IBlockTree blockTree, IAccountStateProvider stateProvider)
             : this(new ChainHeadSpecProvider(specProvider, blockTree), blockTree, stateProvider)
         {
         }
-        
-        public ChainHeadInfoProvider(IChainHeadSpecProvider specProvider, IBlockTree blockTree, IReadOnlyStateProvider stateProvider)
+
+        public ChainHeadInfoProvider(IChainHeadSpecProvider specProvider, IBlockTree blockTree, IAccountStateProvider stateProvider)
         {
             SpecProvider = specProvider;
-            ReadOnlyStateProvider = stateProvider;
-            _blockTree = blockTree;
+            AccountStateProvider = stateProvider;
+            
+            // TODO: why not NewHeadBlock event?
```

Comment:
This was done this way to be invoked for every added block. To be able to iterate all newly added on chain transactions as well as all removed - reorged ones.